### PR TITLE
[WIP and discussion] using invDXDirect from MSWellHelpers to avoid inverting matrix D

### PR DESCRIPTION
**Pull request by GitPaean**
for StandardWell. Basically, as @blattms pointed out for more than one time, the `invert()` in DUNE is not very stable. This PR using a direct solver to avoid inverting the matrix D for the StandardWell. 

It is for testing and discussion purpose and it is not in a good form for going in.  When we decide what we want, I will revise and make it in a good form to go in. 

Please do not merge unless requested and agreed.

### CODE REVIEW

**Comment by GitPaean** [comment]
jenkins build this pleasse

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
One thing worth noticing is that, with this PR in its current form, `umfpack` will be a hard dependence of `opm-simulators`.

**Comment by akva2** [comment]
if that is the case, update opm-simulators-prereqs.cmake

**Comment by akva2** [comment]
sure, that's what i meant.

**Comment by GitPaean** [comment]
Only test failure with jenkins build is the example of polymer_oilwater case. 

The well curve looks okay. 

image

**Comment by GitPaean** [comment]
The impact on Norne running is negligible. 

With this PR. 
```
61229 ================    End of simulation     ===============
61230 
61231 Total time (seconds):         1159.2
61232 Solver time (seconds):        1154.68
61233  Assembly time (seconds):     632.172 (Failed: 21.4647; 3.39539%)
61234  Linear solve time (seconds): 489.803 (Failed: 19.3961; 3.95998%)
61235  Update time (seconds):       19.7938 (Failed: 0.698609; 3.52943%)
61236  Output write time (seconds): 5.79621
61237 Overall Well Iterations:      977 (Failed: 21; 2.14944%)
61238 Overall Linearizations:       1941 (Failed: 66; 3.40031%)
61239 Overall Newton Iterations:    1602 (Failed: 66; 4.11985%)
61240 Overall Linear Iterations:    24511 (Failed: 966; 3.94109%)
61241 
```

With master branch
```
61242 Total time (seconds):         1158.44
61243 Solver time (seconds):        1154
61244  Assembly time (seconds):     667.213 (Failed: 22.9794; 3.44408%)
61245  Linear solve time (seconds): 455.397 (Failed: 16.9395; 3.71972%)
61246  Update time (seconds):       19.1142 (Failed: 0.711832; 3.7241%)
61247  Output write time (seconds): 5.73246
61248 Overall Well Iterations:      977 (Failed: 21; 2.14944%)
61249 Overall Linearizations:       1948 (Failed: 66; 3.38809%)
61250 Overall Newton Iterations:    1609 (Failed: 66; 4.10193%)
61251 Overall Linear Iterations:    24455 (Failed: 894; 3.65569%)
```

**Comment by atgeirr** [comment]
Looks promising: after checking model 2, please make this ready for inclusion. In particular, the invDuneD_ member is now no longer an inverse (great!).

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
Not sure why after revising the PR, one more test failed, the SPE1CASE2_THERMAL. 
The well curves look okay though. 

image

**Comment by GitPaean** [comment]
I got mixed message with model 2 running. The performance fluctuation is quite significant. 
The following is two running result with this PR.  It looks like it increased the iterations for a 1-2%, naturally, it might impact the performance of model 2 very slightly. 
run 1
```
171167 Total time (seconds):         4975.79
171168 Solver time (seconds):        4958.63
171169  Assembly time (seconds):     1953.31 (Failed: 156.01; 7.98692%)
171170  Linear solve time (seconds): 2907.89 (Failed: 183.84; 6.32213%)
171171  Update time (seconds):       70.5183 (Failed: 5.19954; 7.37331%)
171172  Output write time (seconds): 20.1964
171173 Overall Well Iterations:      768 (Failed: 40; 5.20833%)
171174 Overall Linearizations:       2225 (Failed: 176; 7.91011%)
171175 Overall Newton Iterations:    1888 (Failed: 176; 9.32203%)
171176 Overall Linear Iterations:    47344 (Failed: 2439; 5.15166%)
```
run 2
```
171167 Total time (seconds):         4356.49
171168 Solver time (seconds):        4339.91
171169  Assembly time (seconds):     1910.74 (Failed: 155.177; 8.1213%)
171170  Linear solve time (seconds): 2346.06 (Failed: 131.741; 5.61541%)
171171  Update time (seconds):       58.1603 (Failed: 4.99557; 8.58932%)
171172  Output write time (seconds): 19.5595
171173 Overall Well Iterations:      768 (Failed: 40; 5.20833%)
171174 Overall Linearizations:       2225 (Failed: 176; 7.91011%)
171175 Overall Newton Iterations:    1888 (Failed: 176; 9.32203%)
171176 Overall Linear Iterations:    47344 (Failed: 2439; 5.15166%)
```

with master branch
```
171061 Total time (seconds):         4667
171062 Solver time (seconds):        4650.77
171063  Assembly time (seconds):     1819.08 (Failed: 131.161; 7.21026%)
171064  Linear solve time (seconds): 2740.93 (Failed: 143.98; 5.25296%)
171065  Update time (seconds):       66.4037 (Failed: 4.31462; 6.49756%)
171066  Output write time (seconds): 19.0483
171067 Overall Well Iterations:      752 (Failed: 38; 5.05319%)
171068 Overall Linearizations:       2156 (Failed: 153; 7.09647%)
171069 Overall Newton Iterations:    1822 (Failed: 153; 8.39737%)
171070 Overall Linear Iterations:    46989 (Failed: 2248; 4.7841%)
```

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by GitPaean** [comment]
jenkins build this opm-tests=15 please

**Comment by GitPaean** [comment]
A new running of model 2. 
With this PR
```
171168 Total time (seconds):         4277.69
171169 Solver time (seconds):        4262.27
171170  Assembly time (seconds):     1844.28 (Failed: 149.713; 8.11773%)
171171  Linear solve time (seconds): 2337.04 (Failed: 130.333; 5.57683%)
171172  Update time (seconds):       57.1016 (Failed: 4.91228; 8.6027%)
171173  Output write time (seconds): 18.187
171174 Overall Well Iterations:      768 (Failed: 40; 5.20833%)
171175 Overall Linearizations:       2225 (Failed: 176; 7.91011%)
171176 Overall Newton Iterations:    1888 (Failed: 176; 9.32203%)
171177 Overall Linear Iterations:    47344 (Failed: 2439; 5.15166%)
```

With master branch
```
171061 
171062 Total time (seconds):         4204.36
171063 Solver time (seconds):        4188.71
171064  Assembly time (seconds):     1771.03 (Failed: 128.831; 7.27434%)
171065  Linear solve time (seconds): 2339.28 (Failed: 120.102; 5.13413%)
171066  Update time (seconds):       54.7167 (Failed: 4.19238; 7.66197%)
171067  Output write time (seconds): 18.3463
171068 Overall Well Iterations:      752 (Failed: 38; 5.05319%)
171069 Overall Linearizations:       2156 (Failed: 153; 7.09647%)
171070 Overall Newton Iterations:    1822 (Failed: 153; 8.39737%)
171071 Overall Linear Iterations:    46989 (Failed: 2248; 4.7841%)
```

**Comment by GitPaean** [comment]
With this PR, we might see 1-2% performance degeneration, mostly from the changed iteration behavoir, which is difficult to control. It is basically the correct direction to go to avoid future matrix inverting problem. 

We might be able to improve slightly the performance introduced by using linear solution, but it is not very clear how now. 

To me, this PR is ready to go in.

**Comment by GitPaean** [comment]
jenkins build this opm-tests=15 please

**Comment by GitPaean** [comment]
jenkins build this opm-tests=15 please

**Comment by GitPaean** [comment]
I do not understand the failure. Something related to opm-tests. I force pushed the last commit with some modification. 

```
~/workspace/opm-simulators-PR-builder ~/workspace/opm-simulators-PR-builder
Initialized empty Git repository in /var/lib/jenkins/workspace/opm-simulators-PR-builder/deps/opm-tests/.git/
From https://github.com/OPM/opm-tests
 * branch            master     -> FETCH_HEAD
 * [new branch]      master     -> origin/master
From https://github.com/OPM/opm-tests
 * [new ref]         refs/pull/15/merge -> branch_to_build
Already on 'master'
Branch master set up to track remote branch master from origin.
fatal: refusing to merge unrelated histories
Build step 'Execute shell' marked build as failure
Skipped archiving because build is not successful
Recording test results
```

**Comment by akva2** [comment]
your branch is not rebased on master, simple as that. so it is not mergeable.

**Comment by GitPaean** [comment]
> your branch is not rebased on master, simple as that. so it is not mergeable.

Is this a new policy?  because there is no merging conflicts. 
If it is something to be enforced, I support it.

**Comment by GitPaean** [comment]
jenkins build this opm-tests=15 please

**Comment by akva2** [comment]
```
Checking out files: 100% (492/492), done.
Branch master set up to track remote branch master from origin.
Already on 'master'
warning: Cannot merge binary files: spe9group/opm-simulation-reference/flow/SPE9_CP_GROUP.UNRST (HEAD vs. branch_to_build)
warning: Cannot merge binary files: spe9/opm-simulation-reference/flow/SPE9_CP_SHORT.UNSMRY (HEAD vs. branch_to_build)
warning: Cannot merge binary files: spe9/opm-simulation-reference/flow/SPE9_CP_SHORT.UNRST (HEAD vs. branch_to_build)
warning: Cannot merge binary files: spe5/opm-simulation-reference/flow/SPE5CASE1.UNSMRY (HEAD vs. branch_to_build)
warning: Cannot merge binary files: spe5/opm-simulation-reference/flow/SPE5CASE1.UNRST (HEAD vs. branch_to_build)
warning: Cannot merge binary files: spe3/opm-simulation-reference/flow/SPE3CASE1.UNSMRY (HEAD vs. branch_to_build)
warning: Cannot merge binary files: spe3/opm-simulation-reference/flow/SPE3CASE1.UNRST (HEAD vs. branch_to_build)
warning: Cannot merge binary files: spe1/opm-simulation-reference/flow/SPE1CASE2_THERMAL.UNSMRY (HEAD vs. branch_to_build)
warning: Cannot merge binary files: spe1/opm-simulation-reference/flow/SPE1CASE2_THERMAL.UNRST (HEAD vs. branch_to_build)
warning: Cannot merge binary files: spe1/opm-simulation-reference/flow/SPE1CASE2_OILGAS.UNSMRY (HEAD vs. branch_to_build)
warning: Cannot merge binary files: spe1/opm-simulation-reference/flow/SPE1CASE2_OILGAS.UNRST (HEAD vs. branch_to_build)
warning: Cannot merge binary files: spe1/opm-simulation-reference/flow/SPE1CASE2_2P.UNRST (HEAD vs. branch_to_build)
warning: Cannot merge binary files: spe1/opm-simulation-reference/flow/SPE1CASE2.UNSMRY (HEAD vs. branch_to_build)
warning: Cannot merge binary files: spe1/opm-simulation-reference/flow/SPE1CASE2.UNRST (HEAD vs. branch_to_build)
warning: Cannot merge binary files: polymer_simple2D/opm-simulation-reference/flow/2D_THREEPHASE_POLY_HETER.UNSMRY (HEAD vs. branch_to_build)
warning: Cannot merge binary files: polymer_simple2D/opm-simulation-reference/flow/2D_THREEPHASE_POLY_HETER.UNRST (HEAD vs. branch_to_build)
warning: Cannot merge binary files: polymer_oilwater/opm-simulation-reference/flow/2D_OILWATER_POLYMER.UNSMRY (HEAD vs. branch_to_build)
warning: Cannot merge binary files: polymer_oilwater/opm-simulation-reference/flow/2D_OILWATER_POLYMER.UNRST (HEAD vs. branch_to_build)
warning: Cannot merge binary files: msw_3d_hfa/opm-simulation-reference/flow/3D_MSW.UNSMRY (HEAD vs. branch_to_build)
warning: Cannot merge binary files: msw_3d_hfa/opm-simulation-reference/flow/3D_MSW.UNRST (HEAD vs. branch_to_build)
warning: Cannot merge binary files: msw_2d_h/opm-simulation-reference/flow/2D_H__.UNSMRY (HEAD vs. branch_to_build)
warning: Cannot merge binary files: msw_2d_h/opm-simulation-reference/flow/2D_H__.UNRST (HEAD vs. branch_to_build)
Auto-merging spe9group/opm-simulation-reference/flow/SPE9_CP_GROUP.UNRST
CONFLICT (add/add): Merge conflict in spe9group/opm-simulation-reference/flow/SPE9_CP_GROUP.UNRST
Auto-merging spe9group/opm-simulation-reference/flow/SPE9_CP_GROUP.PRT
CONFLICT (add/add): Merge conflict in spe9group/opm-simulation-reference/flow/SPE9_CP_GROUP.PRT
Auto-merging spe9/opm-simulation-reference/flow_legacy/SPE9_CP_SHORT.PRT
CONFLICT (add/add): Merge conflict in spe9/opm-simulation-reference/flow_legacy/SPE9_CP_SHORT.PRT
Auto-merging spe9/opm-simulation-reference/flow/SPE9_CP_SHORT.UNSMRY
CONFLICT (add/add): Merge conflict in spe9/opm-simulation-reference/flow/SPE9_CP_SHORT.UNSMRY
Auto-merging spe9/opm-simulation-reference/flow/SPE9_CP_SHORT.UNRST
CONFLICT (add/add): Merge conflict in spe9/opm-simulation-reference/flow/SPE9_CP_SHORT.UNRST
Auto-merging spe9/opm-simulation-reference/flow/SPE9_CP_SHORT.PRT
CONFLICT (add/add): Merge conflict in spe9/opm-simulation-reference/flow/SPE9_CP_SHORT.PRT
Auto-merging spe5/opm-simulation-reference/flow/SPE5CASE1.UNSMRY
CONFLICT (add/add): Merge conflict in spe5/opm-simulation-reference/flow/SPE5CASE1.UNSMRY
Auto-merging spe5/opm-simulation-reference/flow/SPE5CASE1.UNRST
CONFLICT (add/add): Merge conflict in spe5/opm-simulation-reference/flow/SPE5CASE1.UNRST
Auto-merging spe3/opm-simulation-reference/flow_legacy/SPE3CASE1.PRT
CONFLICT (add/add): Merge conflict in spe3/opm-simulation-reference/flow_legacy/SPE3CASE1.PRT
Auto-merging spe3/opm-simulation-reference/flow/SPE3CASE1.UNSMRY
CONFLICT (add/add): Merge conflict in spe3/opm-simulation-reference/flow/SPE3CASE1.UNSMRY
Auto-merging spe3/opm-simulation-reference/flow/SPE3CASE1.UNRST
CONFLICT (add/add): Merge conflict in spe3/opm-simulation-reference/flow/SPE3CASE1.UNRST
Auto-merging spe3/opm-simulation-reference/flow/SPE3CASE1.PRT
CONFLICT (add/add): Merge conflict in spe3/opm-simulation-reference/flow/SPE3CASE1.PRT
Auto-merging spe1/opm-simulation-reference/flow/SPE1CASE2_THERMAL.UNSMRY
CONFLICT (add/add): Merge conflict in spe1/opm-simulation-reference/flow/SPE1CASE2_THERMAL.UNSMRY
Auto-merging spe1/opm-simulation-reference/flow/SPE1CASE2_THERMAL.UNRST
CONFLICT (add/add): Merge conflict in spe1/opm-simulation-reference/flow/SPE1CASE2_THERMAL.UNRST
Auto-merging spe1/opm-simulation-reference/flow/SPE1CASE2_OILGAS.UNSMRY
CONFLICT (add/add): Merge conflict in spe1/opm-simulation-reference/flow/SPE1CASE2_OILGAS.UNSMRY
Auto-merging spe1/opm-simulation-reference/flow/SPE1CASE2_OILGAS.UNRST
CONFLICT (add/add): Merge conflict in spe1/opm-simulation-reference/flow/SPE1CASE2_OILGAS.UNRST
Auto-merging spe1/opm-simulation-reference/flow/SPE1CASE2_2P.UNRST
CONFLICT (add/add): Merge conflict in spe1/opm-simulation-reference/flow/SPE1CASE2_2P.UNRST
Auto-merging spe1/opm-simulation-reference/flow/SPE1CASE2.UNSMRY
CONFLICT (add/add): Merge conflict in spe1/opm-simulation-reference/flow/SPE1CASE2.UNSMRY
Auto-merging spe1/opm-simulation-reference/flow/SPE1CASE2.UNRST
CONFLICT (add/add): Merge conflict in spe1/opm-simulation-reference/flow/SPE1CASE2.UNRST
Auto-merging polymer_simple2D/opm-simulation-reference/flow/2D_THREEPHASE_POLY_HETER.UNSMRY
CONFLICT (add/add): Merge conflict in polymer_simple2D/opm-simulation-reference/flow/2D_THREEPHASE_POLY_HETER.UNSMRY
Auto-merging polymer_simple2D/opm-simulation-reference/flow/2D_THREEPHASE_POLY_HETER.UNRST
CONFLICT (add/add): Merge conflict in polymer_simple2D/opm-simulation-reference/flow/2D_THREEPHASE_POLY_HETER.UNRST
Auto-merging polymer_oilwater/opm-simulation-reference/flow/2D_OILWATER_POLYMER.UNSMRY
CONFLICT (add/add): Merge conflict in polymer_oilwater/opm-simulation-reference/flow/2D_OILWATER_POLYMER.UNSMRY
Auto-merging polymer_oilwater/opm-simulation-reference/flow/2D_OILWATER_POLYMER.UNRST
CONFLICT (add/add): Merge conflict in polymer_oilwater/opm-simulation-reference/flow/2D_OILWATER_POLYMER.UNRST
Auto-merging msw_3d_hfa/opm-simulation-reference/flow/3D_MSW.UNSMRY
CONFLICT (add/add): Merge conflict in msw_3d_hfa/opm-simulation-reference/flow/3D_MSW.UNSMRY
Auto-merging msw_3d_hfa/opm-simulation-reference/flow/3D_MSW.UNRST
CONFLICT (add/add): Merge conflict in msw_3d_hfa/opm-simulation-reference/flow/3D_MSW.UNRST
Auto-merging msw_2d_h/opm-simulation-reference/flow/2D_H__.UNSMRY
CONFLICT (add/add): Merge conflict in msw_2d_h/opm-simulation-reference/flow/2D_H__.UNSMRY
Auto-merging msw_2d_h/opm-simulation-reference/flow/2D_H__.UNRST
CONFLICT (add/add): Merge conflict in msw_2d_h/opm-simulation-reference/flow/2D_H__.UNRST
Automatic merge failed; fix conflicts and then commit the result.
```
lots of conflicts, just not shown on jenkins for some reason.

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by GitPaean** [comment]
OPM/opm-tests#16 was created manually based on the master branch, and OPM/opm-tests#15 was created with `update_data`, both got the following problem.  I just tested both PRs locally, I did not see any merging conflict. 

```Already on 'master'
Branch master set up to track remote branch master from origin.
fatal: refusing to merge unrelated histories
Build step 'Execute shell' marked build as failure
Skipped archiving because build is not successful
Recording test results
ERROR: Step ?Publish JUnit test result report? failed: No test report files were found. Configuration error?```

**Comment by GitPaean** [comment]
jenkins build this opm-tests=16 please

**Comment by akva2** [comment]
something must have broke with the recent opm-common merge. you are spamming builds so i can't really look into what's going on before it settles.

**Comment by akva2** [comment]
jenkins build this opm-tests=16 opm-common=405 please

**Comment by akva2** [comment]
jenkins build this opm-tests=16 opm-common=405 please

**Comment by akva2** [comment]
there u go, sorry for inconvenience.

**Comment by GitPaean** [comment]
> there u go, sorry for inconvenience.

Thanks for the help.

As I said above, this PR should go in to avoid future dune inverting failure.

**Comment by GitPaean** [comment]
> I.e. set up UMFPack up front (once D is defined) and only call its apply method in Wells::apply

If the decomposition happens when we create the UMFPack, then it is for sure will be a more optimal solution. I just think the overhead should be very small since it is a so small matrix and did not prioritize this without the knowledge. I will revise the PR following your suggestion.

**Comment by GitPaean** [comment]
jenkins build this opm-tests=16 opm-common=405 please

**Comment by GitPaean** [comment]
@blattms , the commit 904133f7df8e1977a6abc6cd2f798935757307d3 does not work. Could you please shed some light about it?

**Comment by GitPaean** [comment]
@blattms  I tried to reuse the `UMFPack`, while it ran into some umfapck and segmentation fault problem. 

It is in the commit 904133f7df8e1977a6abc6cd2f798935757307d3, I am not sure you have any idea about the problem?

**Comment by GitPaean** [comment]
I tried to fix https://github.com/GitPaean/opm-simulators/tree/using_lu_invert_2 , while without good news. 

I totally agree with @blattms , the current way is suboptimal, it might impact the performance a little bit (especially the multisegment well part). It is hard to prioritize more time to improve the current implementation. I will suggest we merge this PR and hopefully, people with more experience in DUNE can help to provide a better way to reuse the decomposition.

File: opm/autodiff/StandardWell_impl.hpp [hunk 1]
```diff
@@ -2198,7 +2188,7 @@ namespace Opm
 
                 Dune::FieldMatrix<Scalar, numWellEq, numEq> tmp;
                 typename Mat::block_type tmp1;
-                Dune::FMatrixHelp::multMatrix(invDuneD_[0][0],  (*colB), tmp);
+                Dune::FMatrixHelp::multMatrix(duneD_[0][0],  (*colB), tmp);
```
Comment (GitPaean) [inline opm/autodiff/StandardWell_impl.hpp]: Since we do not have `invDuneD_` anymore, this code will be impacted.

Comment (GitPaean) [reply opm/autodiff/StandardWell_impl.hpp]: Maybe
```
duneD_[0][0].solve(tmp, (*colB));
```

Comment (GitPaean) [reply opm/autodiff/StandardWell_impl.hpp]: (*colB) is a FieldMatrix, it is not right!

**Comment by blattms** [comment]
My only wild guess about the problem with https://github.com/GitPaean/opm-simulators/tree/using_lu_invert_2 is that the assignment is not working as expected. I would blame their use of working space arrays. You could tried to
- use setMatrix(D) instead of assigning a new UMFPack, or 
- using a smart pointer for UMFPack and use reset(new UMFPack(...)) instead of assignment. 
If either works then we should delete the copy constructor upstream.

Could you post the error message you got?

**Comment by GitPaean** [comment]
> Could you post the error message you got?

```
Thread 1 "flow" received signal SIGSEGV, Segmentation fault.
0x00007ffff7527909 in umfdi_solve () from /usr/lib/x86_64-linux-gnu/libumfpack.so.5.7.1

(gdb) backtrace 
#0  0x00007ffff7527909 in umfdi_solve () from /usr/lib/x86_64-linux-gnu/libumfpack.so.5.7.1
#1  0x00007ffff75375f2 in umfpack_di_solve () from /usr/lib/x86_64-linux-gnu/libumfpack.so.5.7.1
#2  0x000000000078f8ef in Dune::UMFPackMethodChooser<double>::solve<int, int*, int*, double*, double*, double*, void*, double*, double*> () at /home/kaib/local/dune2.5.2/release2.5.2_O3_mpi/include/dune/istl/umfpack.hh:108
#3  Dune::UMFPack<Dune::BCRSMatrix<Dune::FieldMatrix<double, 3, 3>, std::allocator<Dune::FieldMatrix<double, 3, 3> > > >::apply (this=0x1b081e0, x=..., b=..., res=...)
    at /home/kaib/local/dune2.5.2/release2.5.2_O3_mpi/include/dune/istl/umfpack.hh:329
#4  0x00000000008f4a2c in Opm::wellhelpers::invDXDirect<Dune::UMFPack<Dune::BCRSMatrix<Dune::FieldMatrix<double, 3, 3>, std::allocator<Dune::FieldMatrix<double, 3, 3> > > >, Dune::BlockVector<Dune::FieldVector<double, 3>, std::allocator<Dune::FieldVector<double, 3> > > > (linsolver=..., x=...) at /home/kaib/OPM-PR-test/debug/opm-simulators/opm/autodiff/WellHelpers.hpp:209
#5  0x00000000009014ed in Opm::StandardWell<Ewoms::Properties::TTag::EclFlowProblem>::solveEqAndUpdateWellState (this=0x1b07eb0, well_state=...)
    at /home/kaib/OPM-PR-test/debug/opm-simulators/opm/autodiff/StandardWell_impl.hpp:1618
#6  0x000000000090c2b4 in Opm::BlackoilWellModel<Ewoms::Properties::TTag::EclFlowProblem>::solveWellEq (this=this@entry=0x7fffffffce00, dt=dt@entry=86400)
    at /home/kaib/OPM-PR-test/debug/opm-simulators/opm/autodiff/BlackoilWellModel_impl.hpp:494
#7  0x000000000090e153 in Opm::BlackoilWellModel<Ewoms::Properties::TTag::EclFlowProblem>::assemble (this=0x7fffffffce00, iterationIdx=iterationIdx@entry=0, dt=86400)
    at /home/kaib/OPM-PR-test/debug/opm-simulators/opm/autodiff/BlackoilWellModel_impl.hpp:281
#8  0x000000000092eb49 in Opm::BlackoilModelEbos<Ewoms::Properties::TTag::EclFlowProblem>::assemble (iterationIdx=0, timer=..., this=0x1b0abb0) at /home/kaib/OPM-PR-test/debug/opm-simulators/opm/autodiff/BlackoilModelEbos.hpp:376
#9  Opm::BlackoilModelEbos<Ewoms::Properties::TTag::EclFlowProblem>::nonlinearIteration<Opm::NonlinearSolver<Opm::BlackoilModelEbos<Ewoms::Properties::TTag::EclFlowProblem> > > (nonlinear_solver=..., timer=..., iteration=0, 
    this=0x1b0abb0) at /home/kaib/OPM-PR-test/debug/opm-simulators/opm/autodiff/BlackoilModelEbos.hpp:248
#10 Opm::NonlinearSolver<Opm::BlackoilModelEbos<Ewoms::Properties::TTag::EclFlowProblem> >::step (this=this@entry=0x1a6b1d0, timer=..., initial_reservoir_state=..., initial_well_state=..., reservoir_state=..., well_state=...)
    at /home/kaib/OPM-PR-test/debug/opm-simulators/opm/autodiff/NonlinearSolver_impl.hpp:145
#11 0x0000000000931d95 in Opm::NonlinearSolver<Opm::BlackoilModelEbos<Ewoms::Properties::TTag::EclFlowProblem> >::step (well_state=..., reservoir_state=..., timer=..., this=0x1a6b1d0)
    at /home/kaib/OPM-PR-test/debug/opm-simulators/opm/autodiff/NonlinearSolver_impl.hpp:111
#12 Opm::AdaptiveTimeStepping::stepImpl<Opm::NonlinearSolver<Opm::BlackoilModelEbos<Ewoms::Properties::TTag::EclFlowProblem> >, Opm::BlackoilState, Opm::WellStateFullyImplicitBlackoil, Opm::BlackoilOutputEbos<Ewoms::Properties::TTag::EclFlowProblem> > (this=this@entry=0x1b0a640, simulatorTimer=..., solver=..., state=..., well_state=..., event=true, outputWriter=0x1a6e6e0, fipnum=0x0)
    at /home/kaib/OPM-PR-test/debug/opm-simulators/opm/simulators/timestepping/AdaptiveTimeStepping_impl.hpp:236
#13 0x000000000094ce43 in Opm::AdaptiveTimeStepping::step<Opm::NonlinearSolver<Opm::BlackoilModelEbos<Ewoms::Properties::TTag::EclFlowProblem> >, Opm::BlackoilState, Opm::WellStateFullyImplicitBlackoil, Opm::BlackoilOutputEbos<Ewoms::Properties::TTag::EclFlowProblem> > (fipnum=0x0, outputWriter=..., event=<optimized out>, well_state=..., state=..., solver=..., simulatorTimer=..., this=<optimized out>)
    at /home/kaib/OPM-PR-test/debug/opm-simulators/opm/simulators/timestepping/AdaptiveTimeStepping_impl.hpp:177
#14 Opm::SimulatorFullyImplicitBlackoilEbos<Ewoms::Properties::TTag::EclFlowProblem>::run (this=0x1b09b00, timer=...) at /home/kaib/OPM-PR-test/debug/opm-simulators/opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp:250
#15 0x000000000094ee8b in Opm::FlowMainEbos<Ewoms::Properties::TTag::EclFlowProblem>::runSimulator (this=this@entry=0x7fffffffdc10) at /home/kaib/OPM-PR-test/debug/opm-simulators/opm/autodiff/FlowMainEbos.hpp:517
#16 0x00000000009537be in Opm::FlowMainEbos<Ewoms::Properties::TTag::EclFlowProblem>::execute (this=this@entry=0x7fffffffdc10, argc=<optimized out>, argv=<optimized out>)
    at /home/kaib/OPM-PR-test/debug/opm-simulators/opm/autodiff/FlowMainEbos.hpp:115
#17 0x0000000000786fcd in Opm::flowEbosBlackoilMain (argc=3, argv=0x7fffffffe348) at /home/kaib/OPM-PR-test/debug/opm-simulators/opm/simulators/flow_ebos_blackoil.cpp:59
#18 0x0000000000747f47 in main (argc=3, argv=0x7fffffffe348) at /home/kaib/OPM-PR-test/debug/opm-simulators/examples/flow.cpp:200
```

**Comment by GitPaean** [comment]
> is that the assignment is not working as expected.

Most likely it is true. 

> using a smart pointer for UMFPack and use reset(new UMFPack(...)) instead of assignment.
If either works then we should delete the copy constructor upstream.

I tried this one, it is basically the same. 

It is a memory problem, for most of the models, it stops working from the beginning, for some model, it runs a few time steps, then segmentation fault. It runs through some cases with valgrind on. 

> use setMatrix(D) instead of assigning a new UMFPack

I can investigate this one when finding time.

**Comment by GitPaean** [comment]
> use setMatrix(D) instead of assigning a new UMFPack

FYI, `setMatrix` works pretty fine.  At least, this should be used to optimize the multisegment well part, which has a more significant matrix to decompose.  For the StandardWell, we need to check if we want to use UMFPack.

**Comment by GitPaean** [comment]
Most likely (not 100% yet), we will not use UMFPack for the StandardWell, while the discussion (mostly related to `setMatrix`) is suggesting to re-use the LU decomposition for the MultisegmentWell. A PR will be proposed in this direction.

**Comment by GitPaean** [comment]
The discussion in this PR related to how to re-use the LU Decomposition will be used to improve the performance of Multisegment wells. For Standard Well, since the matrix D is so small, it is very likely that we will not use UMFPack for this. Closing this PR now.

**Pull request closed without merging**